docs/dev/design/redirects.rst

```diff
@@ -0,0 +1,338 @@
+
+Don't run redirects on domains from pull request previews
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+We currently run redirects on domains from pull request previews,
+this is a problem when moving a whole project to a new domain.
+
+We don't the need to run redirects on external domains, they
+should be treated as temporary domains.
+
+Normalize paths with trailing slashes
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Currently, if users want to redirect a path with a trailing slash and without it,
+they need to create two separate redirects (``/page/`` and ``/page``).
+
+We can simplify this by normalizing the path before matching it.
+
+For example:
+
+From:
+  ``/page/``
+To:
+  ``/new/page``
+
+The from path will be normalized to ``/page``,
+and the filename to match will also be normalized before matching it.
+This is similar to what Netlify does:
+https://docs.netlify.com/routing/redirects/redirect-options/#trailing-slash.
+
+Page and exact redirects without a wildcard at the end will be normalized,
+all other redirects need to be matched as is.
+
+This makes it impossible to match a path with a trailing slash.
+
+Explicit ``$rest`` placeholder
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Explicitly place the ``$rest`` placeholder in the target URL,
+instead of adding it automatically.
+
+Some times users want to redirect to a different path,
+we have been adding a query parameter in the target URL to
+prevent the old path from being added in the final path.
+For example ``/new/path/?_=``.
+
+Instead of adding the path automatically,
+users have to add the ``$rest`` placeholder in the target URL.
+For example:
+
+From:
+  ``/old/path/$rest``
+To:
+  ``/new/path/$rest``
+
+From:
+  ``/old/path/$rest``
+To:
+  ``/new/path/?page=$rest&foo=bar``
+
+Improving page redirects
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+- Allow to redirect to external domains.
+  This can be useful to apply a redirect of a well known path
+  in all versions to another domain.
+
+  For example, ``/security/`` to a their security policy page in another domain.
+
+  This new feature isn't strictly needed,
+  but it will be useful to simplify the explanation of the feature
+  (one less restriction to explain).
+
+- Allow a wildcard at the end of the from path.
+  This will allow users to migrate a whole directory to a new path
+  without having to create an exact redirect for each version.
+
+  Similar to exact redirects, users need to add the ``$rest`` placeholder explicitly.
+  This means that that page redirects are the same as exact redirects,
+  with the only difference that they apply to all versions.
+
+Improving Sphinx redirects
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+These redirects are useful, but we should rename them to something more general,
+since they apply to all types of projects, not just Sphinx projects.
+
+Proposed names:
+
+- HTML URL to clean URL redirect (``file.html`` to ``file/``)
+- Clean URL to HTML URL redirect (``file/`` to ``file.html``)
```

Comment:
I've seen this named "Fancy URLs" in other doctools.

Reply:
Wikipedia has several other names https://en.wikipedia.org/wiki/Clean_URL

> Clean URLs (also known as user-friendly URLs, pretty URLs, search engine-friendly URLs or RESTful URLs)

Reply:
Oh, yes. These two are concepts that I've already read as well: "user-friendly URLs, pretty URLs"

Reply:
We can probably link to that wikipedia page from our glossary 😉